File: app/Controllers/ChapaController.hs
module Controllers.ChapaController where

import Control.Exception
import Database.PostgreSQL.Simple
import Models.Chapa
import Models.Admin
import Models.Estudante
import Models.Votacao
import Control.Exception

adicionaVotoChapa :: Connection -> Int -> IO ()
adicionaVotoChapa conn idChapa = do
  adicionaVoto conn idChapa

getChapas :: Connection -> IO [ChapaVisualization]
getChapas conn = getChapasVotacaoAtiva conn

cadastraEstudanteEmChapa :: Connection -> String -> String -> String -> Int -> String -> IO ()
cadastraEstudanteEmChapa conn loginAdmin senhaAdmin matricula idChapa diretoria = do
  resultadoAdmin <- getAdmin conn loginAdmin senhaAdmin
  if resultadoAdmin /= []
    then do
      resultEstudante <- try (getEstudanteByMatricula conn matricula) :: IO (Either SomeException [Estudante])
      case resultEstudante of
        Left e -> putStrLn $ "Caught exception: " ++ show e
        Right estudante -> do
          if null estudante
            then putStrLn "Estudante não encontrado"
            else do
              resultChapa <- try (getChapaById conn idChapa) :: IO (Either SomeException [Chapa])
              case resultChapa of
                Left e -> putStrLn $ "Caught exception: " ++ show e
                Right chapa -> do
                  if null chapa
                    then putStrLn "Chapa não encontrada"
                  else cadastraEstudanteChapa conn matricula idChapa (idVotacao (head chapa)) diretoria
  else putStrLn "Administrador não está cadastrado no sistema"


removeEstudanteDaChapa :: Connection -> String -> String -> String -> Int -> IO()
removeEstudanteDaChapa conn loginAdmin senhaAdmin matricula idChapa = do
    resultadoAdmin <- getAdmin conn loginAdmin senhaAdmin
    if resultadoAdmin /= []
        then do
            resultEstudante <- try (getEstudanteByMatricula conn matricula) :: IO (Either SomeException [Estudante])
            case resultEstudante of
                Left e -> putStrLn $ "Caught exception: " ++ show e
                Right estudante -> do
                    if null estudante then putStrLn "Estudante não encontrado"
                    else do
                        resultChapa <- try (getChapaById conn idChapa) :: IO (Either SomeException [Chapa])
                        case resultChapa of
                            Left e -> putStrLn $ "Caught exception: " ++ show e
                            Right chapa -> do
                                if null chapa then putStrLn "Chapa não encontrada"
                                else removeEstudanteChapa conn matricula idChapa
    else
        putStrLn "Administrador não está cadastrado no sistema"

cadastraChapa :: Connection -> String -> String -> String -> Int -> Int -> IO ()
cadastraChapa conn loginAdmin senhaAdmin nomeChapa numeroChapa idVotacaoChapa = do
  cadastrarChapa conn loginAdmin senhaAdmin nomeChapa numeroChapa idVotacaoChapa

editaNomeChapa :: Connection -> String -> String -> Int -> String -> IO ()
editaNomeChapa conn login senha idChapa novoNome = do
  editarNomeChapa conn login senha idChapa novoNome

editaNumeroChapa :: Connection -> String -> String -> Int -> Int -> IO ()
editaNumeroChapa conn login senha idChapa novoNumero = do
    editarNumeroChapa conn login senha idChapa novoNumero

removeChapa :: Connection -> String -> String -> Int -> IO ()
removeChapa conn loginAdmin senhaAdmin idChapaRemocao = do
  removerChapa conn loginAdmin senhaAdmin idChapaRemocao

getVotosEmChapas :: Connection -> Int -> IO Int
getVotosEmChapas conn idVotacao = do
  getVotosChapasByVotacao conn idVotacao

getChapaVencedora :: Connection -> Int -> IO String
getChapaVencedora conn idVotacao = do
  chapaVencedora conn idVotacao

getQtdVotosVencedora :: Connection -> Int -> IO Int
getQtdVotosVencedora conn idVotacao = do
  qtdVotosVencedora conn idVotacao

File: logico/Controller/ChapaController.pl
:- include('../Models/Estudante.pl').
:- include('../Models/Chapa.pl').

cadastrar_estudante_chapa(Matricula, Id_chapa, R) :-
    (estudante_cadastrado(Matricula) ->
        (chapa_cadastrada(Id_chapa) -> 
            (not(verifica_estudante_em_chapa(Matricula, Id_chapa)) -> 
                cadastra_estudante_chapa(Matricula, Id_chapa), R = "Estudante cadastrado na chapa";
                R = "Estudante já cadastrado na chapa");
            R = "Chapa não cadastrada");
        R = "Estudante não cadastrado"
    ).

remover_estudante_chapa(Matricula, Id_chapa, R) :-
    (estudante_cadastrado(Matricula) ->
        (chapa_cadastrada(Id_chapa) -> 
            (verifica_estudante_em_chapa(Matricula, Id_chapa) -> 
                remove_estudante_chapa(Matricula, Id_chapa), R = "Estudante removido da chapa";
                R = "Estudante não está cadastrado na chapa");
            R = "Chapa não cadastrada");
        R = "Estudante não cadastrado"
    ).
get_chapas_votacoes_ativas(Result) :- get_chapas_ativas(Result).

adiciona_voto_chapa(ChapaNum, IdVotacao) :- adiciona_voto(ChapaNum, IdVotacao).

verifica_chapa_by_numero_e_votacao(ChapaNumero, IdVotacao) :- verifica_by_numero_votacao(ChapaNumero, IdVotacao).

cadastra_chapa(Nome , Numero, IdVotacao, R) :-
    cadastrar_chapa(Nome, Numero, IdVotacao, R).

edita_nome_chapa(IdChapa, Nome, R) :-
    edita_nome(IdChapa, Nome, R).

edita_numero_chapa(IdChapa, Numero, R) :-
    edita_numero(IdChapa, Numero, R).

get_chapa_vencedora(IDVotacao, ChapaVencedora) :-
    get_chapas_by_votacao(IDVotacao, Chapas),
    get_id_chapa_mais_votos(Chapas, IDChapaVencedora),
    get_chapa_by_id(IDChapaVencedora, ChapaVencedora).

get_id_chapa_mais_votos(Chapas, IDChapaVencedora) :-
    get_mais_votos_rec(Chapas, -1, -1, IDChapaVencedora).

get_mais_votos_rec([row(IDChapa, _, _, _, NumVotos)|[]], MaiorQtdVotos, IDMaiorChapaAtual, IDChapaVencedora) :-
    ((NumVotos > MaiorQtdVotos) ->
        IDChapaVencedora = IDChapa;
        IDChapaVencedora = IDMaiorChapaAtual
    ).

get_mais_votos_rec([row(IDChapa, _, _, _, NumVotos) | T], MaiorQtdVotos, IDMaiorChapaAtual, IDChapaVencedora) :-
    ((NumVotos > MaiorQtdVotos) ->
        get_mais_votos_rec(T, NumVotos, IDChapa, IDChapaVencedora);
        get_mais_votos_rec(T, MaiorQtdVotos, IDMaiorChapaAtual, IDChapaVencedora)
    ).

get_votos_chapas_votacao(IdVotacao, Result) :- get_soma_votos_chapas_votacao(IdVotacao, Result).

remover_chapa(Id, R) :-
    (chapa_cadastrada(Id) -> 
        remove_chapa(Id), R = "Chapa removida";
    R = "Chapa não cadastrada").
